jsonschema-hyperschema.xml

```diff
@@ -854,6 +855,87 @@
                     Note that these relationship values are case-insensitive, consistent with their
                     use in HTML and the <xref target="RFC5988">HTTP Link header</xref>.
                 </t>
+                <section title="&quot;collection&quot; and &quot;item&quot; links">
+                    <t>
+                        <xref target="RFC6573">RFC 6573</xref> defines and registers
+                        the "item" and "collection" link relations.  A well-known design pattern
+                        in hypermedia is to use a collection resource to create a member of the
+                        collection and give it a server-assigned URI.  When using HTTP, or a protocol
+                        such as CoAP that is explicitly analogous to HTTP, this is done by POST-ing
+                        a representation of the individual resource to be created to the collection
+                        resource.
+                    </t>
+                    <t>
+                        Resources that are the target of a "collection" link using HTTP or an analogous
+                        protocol in JSON Hyper-Schema MUST NOT assign semantics other than resource
+                        creation to POST (or the analogous method in non-HTTP protocols).
```

Comment:
I'm not sure about this one. Maybe this about what is meant by "resource creation".

Typically, I'm thinking about using the collection+item pattern to describe *relationships*. For instance, consider a `Article tags Tag` relationship model, POST-ing to the URI some article's `tags` relationship (something like `/articles/1/tags/`) may lead to different results depending on the payload and server semantics: it could set a relation between the article and an existing tag (identified by payload) without modifying the underlying entities, or it could add a new tag entity with a relation set. The latter case arguably falls into "resource creation" semantics, but it's not obvious for the former, though one could argue that a new relation is created.

Also, what about a `PUT` or a `DELETE` request? Should we specify the expected behavior as well?

In fact, I wonder if we shouldn't discuss this "protocol semantics" topic separately and let this PR focus on documenting the pattern (where each link should be, how to set the context with `anchorPointer`).

Reply:
@dlax There is no concern with PUT and DELETE, whose semantics are clear.

The problem here is that creating a resource with a server-assigned identifier is not part of HTTP's uniform interface.  But it is one of the most common operations that is done in APIs, and it is flatly impossible to write a generic hypermedia client without it.

Take a look at [application/collection+json](http://amundsen.com/media-types/collection/format/) and [application/maze+xml](http://amundsen.com/media-types/maze/format/), which are both given as the examples of use of "collection" and "item" RFC.  Both of these specify creation-via-POST semantics for collections _based on these link relations_.

For your tag example, there is nothing problematic about any of those behaviors.  The resource has to define what it means to be created.  The point is not to tie "resource creation" to "put a thing in a database", or any other implementation.  Both of the options you give are implementation details.  The point is to ensure that if this _exact_ relation is used, that POSTing to such a resource doesn't do something totally unexpected and damaging, like reboot the application.  Which would otherwise be a completely valid thing for a resource to do in response to a POST (because literally anything is valid for POST).

Reply:
This doesn't even say that creation has to be supported.  It just says not to completely break clients that need *SOME* generic clue as to how to accomplish this.  Link relation types are the only correct way to specify semantics.

And to be clear, these are link/resource semantics, ***NOT*** protocol semantics.  That's kind of the entire point- there is nothing in HTTP that officially carries these semantics.  Nothing we do in this media type changes that.

But per RFC 5988bis:

> Link relation types can also be used to indicate that the target resource has particular attributes, or exhibits particular behaviours

We are following existing patterns, referenced in the RFC that defines these link relation types, to constrain behavior in order to make it possible to implement a generic hypermedia client using JSON Hyper-Schema.

Reply:
@handrews I've thinking about this a bit more...

I think that what made me worry is that the text about "resource creation and POST" is located at the very beginning of the section whereas I would have expected to first read about how to use the collection+item pattern in a JSON Hyper Schema document. (In fact, I'm not that worried and the example I posted was mostly rhetorical.)

So I was about to suggest to move this part (starting from "A well-known design pattern..." to where this comment is attached) later in the section, but I'm actually not sure if it'd make still sense. So if it doesn't, I can live with it as is :)

Reply:
@dlax yeah, that's fair, and I think I can get it to introduce the creation pattern less abruptly.  If it's OK with you I'll just note that for the rewrite, as I'll almost certainly be reworking the text anyway.

Reply:
>  If it's OK with you I'll just note that for the rewrite, as I'll almost certainly be reworking the text anyway.

Sure, that works too.